mono/mini/jit-icalls.c

```diff
@@ -93,8 +94,29 @@ ldvirtfn_internal (MonoObject *obj, MonoMethod *method, gboolean gshared)
 	}
 
 	/* An rgctx wrapper is added by the trampolines no need to do it here */
+	gboolean need_unbox = m_class_is_valuetype (res->klass) && !m_class_is_valuetype (method->klass);
+	if (need_unbox && !mono_use_interpreter) {
```

Comment:
Why interpreter doesn't need unboxing ?

Reply:
Before I fix this issue, I tried interpreter mode. And It worked. So I assume that interpreter was able to handle this without needing unboxing...

Reply:
The interpreter doesn't hit this code path afaik.

Reply:
It seems that it does actually. If I don't add the condition to skip for interpreter, fullaot + interpreter would fail. Maybe it is because of fullaot?

Reply:
Thats possible.

Reply:
This check has no justification behind it, aside that it seems to make stuff work, and looks like a hack. I would like to see the stack trace of the interpreter calling into `mono_ldvirtftn`. Does forcing compilation of the method with jit also fixes the issue ? (`mono_jit_compile_method_jit_only`)

Reply:
It's not a hack. it's because `mono_ldftn` calls `mono_create_jump_trampoline` which does this: https://github.com/mono/mono/blob/fd61732c833e268a1eaa5a2ebf38716921ab3de1/mono/mini/mini-trampolines.c#L1346-L1351

and the call to `ves_icall_create_delegate_ctor` (emitted by the FullAOT compiler) which calls `mono_delegate_ctor` which eventually ends up in `interp_init_delegate` with a delegate that doesn't have `method` or `interp_method` set.

Here's a stack trace from an earlier FullAOT+Interpreter lane:

```
 #7  0x00007ff45583a42a in abort () from /lib/x86_64-linux-gnu/libc.so.6
#8  0x00005618da112a27 in monoeg_assert_abort () at goutput.c:57
#9  0x00005618da0f65b9 in mono_log_write_logfile (log_domain=<optimized out>, level=<optimized out>, hdr=<optimized out>, message=0x7ff43c0051f0 "* Assertion: should not be reached at interp/interp.c:1603\n") at mono-log-common.c:136
#10 0x00005618da11294e in monoeg_g_logstr (msg=<optimized out>, log_level=G_LOG_LEVEL_ERROR, log_domain=0x0) at goutput.c:151
#11 monoeg_g_logv_nofree (log_domain=log_domain@entry=0x0, log_level=log_level@entry=G_LOG_LEVEL_ERROR, format=format@entry=0x5618db1ee568 "* Assertion: should not be reached at %s:%d\n", args=args@entry=0x7ff4492adf08) at goutput.c:166
#12 0x00005618da112ee0 in monoeg_g_logv_nofree (args=0x7ff4492adf08, format=0x5618db1ee568 "* Assertion: should not be reached at %s:%d\n", log_level=G_LOG_LEVEL_ERROR, log_domain=0x0) at goutput.c:159
#13 monoeg_assertion_message (format=format@entry=0x5618db1ee568 "* Assertion: should not be reached at %s:%d\n") at goutput.c:201
#14 0x00005618da112f47 in mono_assertion_message_unreachable (file=file@entry=0x5618db1520e8 "interp/interp.c", line=line@entry=1603) at goutput.c:228
#15 0x00005618d9f4930d in interp_init_delegate (del=0x7ff455485a58, error=0x7ff4492ae080) at interp/interp.c:1603
#16 0x00005618d9e075a0 in mini_init_delegate (delegate=..., error=0x7ff4492ae080) at mini-runtime.c:3764
#17 0x00005618da02ff84 in mono_delegate_ctor_with_method (this_obj=..., target=..., addr=addr@entry=0x7ff452ff6ddb <ut_7811>, method=<optimized out>, error=error@entry=0x7ff4492ae080) at object.c:8767
#18 0x00005618da03013e in mono_delegate_ctor (this_obj=..., this_obj@entry=..., target=..., addr=addr@entry=0x7ff452ff6ddb <ut_7811>, error=error@entry=0x7ff4492ae080) at object.c:8802
#19 0x00005618d9e0d03f in ves_icall_mono_delegate_ctor (this_obj_raw=<optimized out>, target_raw=<optimized out>, addr=0x7ff452ff6ddb <ut_7811>) at jit-icalls.c:1488
#20 0x00007ff4531c22b1 in wrapper_managed_to_native_object___icall_wrapper_ves_icall_mono_delegate_ctor_object_object_intptr () from /mnt/jenkins/workspace/test-mono-pull-request-fullaot+interpreter+llvm/mcs/class/lib/testing_aot_full_interp/mscorlib.dll.so
#21 0x00007ff453031fd2 in System_Collections_Generic_ArraySortHelper_1_T_REF_Sort_T_REF___int_int_System_Collections_Generic_IComparer_1_T_REF () from /mnt/jenkins/workspace/test-mono-pull-request-fullaot+interpreter+llvm/mcs/class/lib/testing_aot_full_interp/mscorlib.dll.so
#22 0x00007ff452a34115 in System_Array_Sort_T_REF_T_REF___int_int_System_Collections_Generic_IComparer_1_T_REF () from /mnt/jenkins/workspace/test-mono-pull-request-fullaot+interpreter+llvm/mcs/class/lib/testing_aot_full_interp/mscorlib.dll.so 
```

I'm not sure where in `interp_create_method_pointer` the pointer gets unboxed, probably some place in `interp_entry_from_trampoline`?

Reply:
I'm missing some background here because I don't know how delegates are related to this change. Only indirectly probably. As far as I know this PR is meant to fix the pattern of `box`, `dup`, `ldvirtftn`, `calli` and my review is based on that. In order to provide some background here are some test cases :

Case1 : https://gist.github.com/BrzVlad/600baa88e1ea4b3259a0023ab7ffbf89 . This case does a normal virtual call and the interpreter does no unboxing. The unboxing still happens on the compiled code side by adding an unbox trampoline to the interp enter wrapper, done by the normal common_call_trampoline.

Case2 : https://gist.github.com/BrzVlad/1fc0511190986c2f04504eacae90d66c . This is basically the same as before but I'm using ldvirtftn rather than doing a normal virtual call. This test case was failing on simple jit before this change, but this PR fixes it. However this 
 is broken in mixed mode, before this PR or with this PR, including the mono_use_interpreter check (that I'm assuming it was supposed to fix it ?). I actually made this test case to see how mixed mode was passing before, without requiring any unboxing, but ldvirtftn is completely broken here. JIT doesn't even get to call into `ldvirtfn_internal` because it crashes somewhere, for me at least ...

To wrap up my point, I'd like to see how interpreter can unbox when it gets called into through `ldvirtftn` + `calli` from compiled code. The fact that some test was passing before this PR, that adds unboxing, is not sufficient evidence for me.  Also `mono_use_interpreter` is overall an ugly variable that should be used as little as possible. It doesn't provide any meaningful information, rather that some methods can be interpreted. How is this helpful in mono_ldvirtftn ? We are called from jitted code and we are resolving a virtual method. This method could end up being either an interpreted method or a jitted method. You are suggesting that calling the interpreter method would not require unboxing, because the interpreter already does it. Even if that is the case, checking for mono_use_interpreter doesn't really tell us wether the resolved method is a jit method or just an interp entry thunk, so we shouldn't consider it when adding an unboxing tramp.